x-pack/plugin/runtime-fields/src/main/java/org/elasticsearch/xpack/runtimefields/mapper/ScriptFieldMapper.java

```diff
@@ -56,81 +73,118 @@ protected String contentType() {
+        }
+
+        @SuppressWarnings("unchecked")
+        static Script parseScript(String name, Object scriptObject) {
+            if (scriptObject instanceof Map) {
+                Map<String, ?> scriptMap = (Map<String, ?>) scriptObject;
+                Object sourceObject = scriptMap.remove("source");
+                if (sourceObject == null) {
+                    throw new IllegalArgumentException("script source must be specified for script field [" + name + "]");
+                }
+                Object langObject = scriptMap.remove("lang");
+                if (langObject != null && langObject.toString().equals(PainlessScriptEngine.NAME) == false) {
+                    throw new IllegalArgumentException("script lang [" + langObject.toString() + "] not supported for script field ["
+                        + name + "]");
+                }
+                Map<String, Object> params;
+                Object paramsObject = scriptMap.remove("params");
+                if (paramsObject != null) {
+                    if (paramsObject instanceof Map == false) {
+                        throw new IllegalArgumentException("unable to parse params for script field [" + name + "]");
+                    }
+                    params = (Map<String, Object>) paramsObject;
+                } else {
+                    params = Collections.emptyMap();
+                }
+                Map<String, String> options;
+                Object optionsObject = scriptMap.remove("options");
+                if (optionsObject != null) {
+                    if (optionsObject instanceof Map == false) {
+                        throw new IllegalArgumentException("unable to parse options for script field [" + name + "]");
+                    }
+                    options = (Map<String, String>) optionsObject;
+                } else {
+                    options = Collections.emptyMap();
+                }
+                if (scriptMap.size() > 0) {
+                    throw new IllegalArgumentException("unsupported parameters specified for script field [" + name + "]: "
+                        + scriptMap.keySet());
+                }
+                return new Script(ScriptType.INLINE, PainlessScriptEngine.NAME, sourceObject.toString(), options, params);
+            } else if (scriptObject instanceof String) {
+                return new Script((String) scriptObject);
+            } else {
+                throw new IllegalArgumentException("unable to parse script for script field [" + name + "]");
+
+            }
         }
     }
 
     public static class TypeParser implements Mapper.TypeParser {
 
-        private final SetOnce<ScriptService> scriptService = new SetOnce<>();
+        // TODO this is quite ugly and it's static which makes it even worse
+        private static final SetOnce<ScriptService> SCRIPT_SERVICE = new SetOnce<>();
 
```

Comment:
Is there no way of getting it via the `ParserContext`?

Reply:
It looks as though we'd have to wire `ScriptService` into `MapperService`, but I think that's a reasonable thing to do

Reply:
I will look into it, thanks for the suggestion.